llama_index/graph_stores/sparql.py

```diff
@@ -0,0 +1,284 @@
+        ?property er:value ?rel1 .
+        ?object er:value ?obj1 .
+                            
+    OPTIONAL {
+        ?triplet2 a er:Triplet ;
+            er:subject ?object ;
+            er:property ?property2 ;
+            er:object ?object2 .
+
+        ?property2 er:value ?rel2 .
+        ?object2 er:value ?obj2 .
+    }}}
+    $limit_str
+        """)
+
+        limit_str = ''
+        if (limit > 0):
+            limit_str = '\nLIMIT ' + str(limit)
+
+        query_string = template.substitute(
+            {'prefixes': self.sparql_prefixes, 'subject': subj,  'limit_str': limit_str})
+
+        triplets_json = self.sparql_query(query_string)
+
+        return self.to_arrows(subj, triplets_json)
+
+    def to_arrows(self, subj: str, rels: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
+        """
+        Convert subject and relations to a string in the desired format.
+        """
+        arrows = []
+        for r in rels:
+            rel1 = r['rel1']['value']
+            obj1 = r['obj1']['value']
+            rel1 = self.unescape_from_rdf(rel1)  # probably unnecessary
+            obj1 = self.unescape_from_rdf(obj1)
+
+            try:
+                rel2 = r['rel2']['value']
+                obj2 = r['obj2']['value']
+                rel2 = self.unescape_from_rdf(rel2)
+                obj2 = self.unescape_from_rdf(obj2)
+                arrows.append(
+                    f"""{subj}, -[{rel1}]->, {obj1}',<-[{rel2}]-, {obj2}""")
+            except:
+                arrows.append(f"""{subj}, -[{rel1}]->, {obj1}""")
+        return {subj: arrows}
+
+# From interface types.py ----------------------------------
+
+# DOES IT ONE-BY-ONE!!!!!!!!!!!!!!!!!!!!!!
+    def upsert_triplet(self, subj: str, rel: str, obj: str) -> None:
+        """Add triplet."""
+        logger.info('sparql upsert_triplet called')
+
+        # crude loop prevention
+        # if obj in self.prior_subjs:
+        #    return
+        # self.prior_subjs.append(subj)
+
+        subj = self.escape_for_rdf(subj)
+        rel = self.escape_for_rdf(rel)
+        obj = self.escape_for_rdf(obj)
+
+        triplet_fragment = '#T'+self.make_id()
+        s_fragment = '#E_'+self.make_id()
+        p_fragment = '#R_'+self.make_id()
+        o_fragment = '#E_'+self.make_id()
+
+        triple = f"""
+            <{triplet_fragment}> a er:Triplet ;
+                        er:subject <{s_fragment}> ;
+                        er:property <{p_fragment}> ;
+                        er:object <{o_fragment}> .
+
+            <{s_fragment}> a er:Entity ;
+                        er:value "{subj}" .
+
+            <{p_fragment}> a er:Relationship ;
+                        er:value "{rel}" .
+
+            <{o_fragment}> a er:Entity ;
+                        er:value "{obj}" .
+                """
+        self.insert_data(triple)
+
+    def client(self) -> Any:
+        """Get client."""
+        logger.info('sparql client(self) called')
+        ...
```

Comment:
Just double-checking, but I'm guessing some of these are still in progress?

Reply:
In progress, yes. I've added type annotations, done a little tidying and just finished tests for sparql.py (and a minimal one for simple.py) I didn't see any tests for the other graph stores...
Is late here now, will check down here properly tomorrow.